### PR TITLE
Difference Equation Transfer Function

### PR DESCRIPTION
**Pull request by EvanMorcom**
### Description

This PR contains the difference equation class that allows for discrete transfer functions to be converted to difference equations.

This will be used to test Python control code and controllers, as well as generate the C code required to run our robot wheel-level controller in firmware.

### Testing Done

1. Manually checked input and output
2. 1st Order step response
3. 2nd Order step response

### Resolved Issues

N/A
### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Comment by EvanMorcom** [comment]
@jonathanlew @garethellis0 @MathewMacDougall @akhilveeraghanta @phelanm20 Trying to break the python code into smaller PR's. I've removed the WIP tag, feel free to review. Thanks :+1:

File: src/firmware/app/control/python_controller/difference_equation.py
Comment (MathewMacDougall) [inline src/firmware/app/control/python_controller/difference_equation.py]: Should `time_step_count` also be "private", ie. `__time_step_count`?

File: src/firmware/app/control/python_controller/difference_equation.py
Comment (MathewMacDougall) [inline src/firmware/app/control/python_controller/difference_equation.py]: `current_input` should (ideally) have type annotation

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

File: environment_setup/setup_software.sh
Comment (garethellis0) [reply environment_setup/setup_software.sh]: We already have a list of software packages above to install, including several python packages (`host_software_packages`). Is there any particular reason we don't just have `python3` in there?

File: src/WORKSPACE
Comment (garethellis0) [inline src/WORKSPACE:195]: Is it possible to move both the `load` of `pip3_import` and and the call to it to the `BUILD` files in `python_controller`?

File: src/firmware/app/control/python_controller/BUILD
Comment (garethellis0) [inline src/firmware/app/control/python_controller/BUILD:28]: This is very neat! :smile:

File: src/firmware/app/control/python_controller/difference_equation.py
Comment (garethellis0) [inline src/firmware/app/control/python_controller/difference_equation.py]: seems like the type (`list`) is duplicated in the doc comment and in the function signature. I think we should be good to remove it from the doc comment?

File: src/firmware/app/control/python_controller/difference_equation.py
Comment (garethellis0) [reply src/firmware/app/control/python_controller/difference_equation.py]: Even if they are, we ought to be consistent, and definitely leave a comment explaining it for non-controls people.

File: src/firmware/app/control/python_controller/difference_equation.py
Comment (garethellis0) [inline src/firmware/app/control/python_controller/difference_equation.py]: I definitely had to look up if this is possible, but could we specify what's contained in the lists? See https://docs.python.org/3/library/typing.html#type-aliases for an example of how to do so.

File: src/firmware/app/control/python_controller/difference_equation.py
Comment (garethellis0) [inline src/firmware/app/control/python_controller/difference_equation.py:17]: After offline discussion with Evan; this class encompasses both the difference equation that will be directly translated to C, and also some simulation stuff for testing that will essentially be duplicated in C. We should split this class into `DifferenceEquation`, which will just contain the actual difference equation, and `DifferenceEquationSimulator` (name tbd), which will hold a `DifferenceEquation` and manage all the history, passing previous values into the difference equation, etc.. This way there's a clear encapsulation for what will be directly translated into generated C code.

File: src/firmware/app/control/python_controller/difference_equation.py
Comment (jonathanlew) [inline src/firmware/app/control/python_controller/difference_equation.py]: Consider using [list comprehensions](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions)

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Looking good!

File: src/firmware/app/control/python_controller/difference_equation.py
Comment (jonathanlew) [inline src/firmware/app/control/python_controller/difference_equation.py]: Consider using filter/map/reduce and list comprehension on lists. Certain for loops are considered [unpythonic](https://docs.quantifiedcode.com/python-anti-patterns/readability/using_an_unpythonic_loop.html)

File: src/firmware/app/control/python_controller/requirements.txt
Comment (jonathanlew) [inline src/firmware/app/control/python_controller/requirements.txt]: Are these being used? Are you leaving these here because we'll use them in the future?

File: src/firmware/app/control/python_controller/difference_equation_test.py
Comment (jonathanlew) [inline src/firmware/app/control/python_controller/difference_equation_test.py]: So we're testing against the python control toolbox? Wondering why we don't use the python control toolbox directly...

**Review by jonathanlew — COMMENTED** [review]
Overall, clean and easy to read

File: environment_setup/setup_software.sh
Comment (EvanMorcom) [reply environment_setup/setup_software.sh]: Good point, I removed this section and added Python3 above

File: src/firmware/app/control/python_controller/difference_equation.py
Comment (EvanMorcom) [reply src/firmware/app/control/python_controller/difference_equation.py]: Done :+1:

File: src/firmware/app/control/python_controller/difference_equation.py
Comment (EvanMorcom) [reply src/firmware/app/control/python_controller/difference_equation.py]: Added a comment specifying the order

File: src/firmware/app/control/python_controller/difference_equation.py
Comment (EvanMorcom) [reply src/firmware/app/control/python_controller/difference_equation.py]: I renamed them both to be 'order'.

Typically 'rank' is used for the denominator as it is the most important in controller stuff, but they are interchangeable.

File: src/firmware/app/control/python_controller/difference_equation.py
Comment (EvanMorcom) [reply src/firmware/app/control/python_controller/difference_equation.py]: Yep, made it 'private' and added a getter function 👍

File: src/firmware/app/control/python_controller/difference_equation.py
Comment (EvanMorcom) [reply src/firmware/app/control/python_controller/difference_equation.py]: Added type :+1:

File: src/WORKSPACE
Comment (EvanMorcom) [reply src/WORKSPACE:195]: Tried copy-pasting all of the pip related commands to the BUILD fail and it didn't compile.

File: src/firmware/app/control/python_controller/difference_equation.py
Comment (EvanMorcom) [reply src/firmware/app/control/python_controller/difference_equation.py]: This is the documentation style that @MathewMacDougall started doing with the python code

File: src/firmware/app/control/python_controller/difference_equation.py
Comment (EvanMorcom) [reply src/firmware/app/control/python_controller/difference_equation.py]: Added list type specification

File: src/firmware/app/control/python_controller/requirements.txt
Comment (EvanMorcom) [reply src/firmware/app/control/python_controller/requirements.txt]: They are actually currently used

File: src/firmware/app/control/python_controller/difference_equation_test.py
Comment (EvanMorcom) [reply src/firmware/app/control/python_controller/difference_equation_test.py]: This is so the python implementation of our control is a close as possible to the C implementation. The discrete transfer functions of the Control toolbox have the exact same output but doesn't exist in a form that we can use in the C firmware

File: src/firmware/app/control/python_controller/requirements.txt
Comment (jonathanlew) [reply src/firmware/app/control/python_controller/requirements.txt]: can you point me to where?

File: src/firmware/app/control/python_controller/difference_equation.py
Comment (MathewMacDougall) [reply src/firmware/app/control/python_controller/difference_equation.py]: oh did I? I agree with gareth lets try just keep the type annotation in the function signature, since that's what actually matters when running it (and comments likely won't stay up to date)

File: src/WORKSPACE
Comment (garethellis0) [reply src/WORKSPACE:195]: confirmed this is the correct usage pattern from the docs: https://github.com/bazelbuild/rules_python/blob/master/docs/pip.md

File: src/firmware/app/control/python_controller/difference_equation_test.py
Comment (akhilveeraghanta) [inline src/firmware/app/control/python_controller/difference_equation_test.py:10]: is there any reason why we chose `unittest` over `pytest`? `pytest` makes things suuuper easy to test, each test is a function instead of a class and `pymock` is magical.

Comment (garethellis0) [reply src/firmware/app/control/python_controller/difference_equation_test.py:10]: Mat, Evan, and I had a chat about this, just seemed like overkill here, no need to add another library unless we have to. At the same time, if it's that much more fully featured, perhaps we can consider switching even if we don't need it here just so that we can be consistent with any future tests that do need those features.

Comment (MathewMacDougall) [reply src/firmware/app/control/python_controller/difference_equation_test.py:10]: Yea I think I mis-remembered / mis-spoke oringinally. I think this is a good opportunity to at least see what it looks like and standardize early.

File: src/firmware/app/control/python_controller/difference_equation.py
Comment (EvanMorcom) [reply src/firmware/app/control/python_controller/difference_equation.py]: Removed the extra type specifications

File: src/firmware/app/control/python_controller/difference_equation.py
Comment (EvanMorcom) [reply src/firmware/app/control/python_controller/difference_equation.py:17]: Broke up into DifferenceEquation and DifferenceEquationSim

File: src/firmware/app/control/python_controller/difference_equation.py
Comment (EvanMorcom) [reply src/firmware/app/control/python_controller/difference_equation.py]: I tried to make this happen but the indices having different values for the lists make this a little complicated with the new code.

Please let me know if you have any suggestions

File: src/firmware/app/control/python_controller/difference_equation.py
Comment (EvanMorcom) [reply src/firmware/app/control/python_controller/difference_equation.py]: See above comment 👍

File: src/firmware/app/control/python_controller/requirements.txt
Comment (EvanMorcom) [reply src/firmware/app/control/python_controller/requirements.txt]: Some of them are hidden inside of other packages.

For example, the python control toolbox uses kiwi solver.

If I don't include kiwi solver here the bazel can't build it

Comment (EvanMorcom) [reply src/firmware/app/control/python_controller/requirements.txt]: But I was able to remove the requirements from the _test.py file 👍

File: src/firmware/app/control/python_controller/difference_equation.py
Comment (EvanMorcom) [reply src/firmware/app/control/python_controller/difference_equation.py]: Done 👍

File: src/firmware/app/control/python_controller/differenceequation.py
Comment (garethellis0) [inline src/firmware/app/control/python_controller/differenceequation.py]: Could we add a return type here?
``` suggestion
    def calculate_output(self, output_history: List[float], input_history: List[float]) -> float:
```

File: src/firmware/app/control/python_controller/requirements.txt
Comment (jonathanlew) [reply src/firmware/app/control/python_controller/requirements.txt]: Oh I see

**Review by jonathanlew — DISMISSED** [review]
LGTM pending other comments

**Comment by EvanMorcom** [comment]
@garethellis0 @MathewMacDougall I got python units tests running in CI finally - please re-review 👍

File: src/firmware/app/control/python_controller/difference_equation_test.py
Comment (MathewMacDougall) [reply src/firmware/app/control/python_controller/difference_equation_test.py:10]: Just for clarity here, @akhilveeraghanta  we're planning to try switch to `pytest` but don't want to block this PR. @EvanMorcom please make a ticket for moving to `pytest`

File: src/firmware/app/control/python_controller/difference_equation_test.py
Comment (MathewMacDougall) [inline src/firmware/app/control/python_controller/difference_equation_test.py]: Just realized this is weird because the file is named `differenceequation.py`. Can we change it to `difference_equation.py` to match our style guide?

File: src/firmware/app/control/python_controller/difference_equation_test.py
Comment (MathewMacDougall) [inline src/firmware/app/control/python_controller/difference_equation_test.py]: what's the `# [s]` comment?

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
just some nits, functionally looks great :+1:

File: src/firmware/app/control/python_controller/difference_equation_test.py
Comment (jonathanlew) [reply src/firmware/app/control/python_controller/difference_equation_test.py:52]: seconds

Comment (jonathanlew) [reply src/firmware/app/control/python_controller/difference_equation_test.py]: s means "in seconds"

File: src/firmware/app/control/python_controller/difference_equation_test.py
Comment (EvanMorcom) [reply src/firmware/app/control/python_controller/difference_equation_test.py:10]: Done, #1381

File: src/firmware/app/control/python_controller/difference_equation_test.py
Comment (EvanMorcom) [reply src/firmware/app/control/python_controller/difference_equation_test.py]: Renamed 👍

File: src/firmware/app/control/python_controller/difference_equation_test.py
Comment (EvanMorcom) [reply src/firmware/app/control/python_controller/difference_equation_test.py]: @jonathanlew is correct, I will rename to [seconds] to be more explicit

**Review by akhilveeraghanta — COMMENTED** [review]
This is super clean :fire: really nice job, just have 1 comment

File: src/firmware/app/control/python_controller/difference_equation.py
Comment (EvanMorcom) [reply src/firmware/app/control/python_controller/difference_equation.py]: Not in this case - I should add a comment.

I should remove the input order almost completely. The input buffer should be the length of the denominator order in case the numerator is not the same order as the denominator

**Review by s-zheng — DISMISSED** [review]

File: src/firmware/app/control/python_controller/difference_equation_test.py
Comment (MathewMacDougall) [inline src/firmware/app/control/python_controller/difference_equation_test.py]: I'm sure `J`, `B`, and `K` have some well-known controls meaning, but could we at least add a quick comment to them? One-letter variable names are really difficult to understand without context. 

Same goes for the `s` variable below (this one seems less special so can probably be renamed more verbosely?)

File: src/firmware/app/control/python_controller/difference_equation_test.py
Comment (MathewMacDougall) [inline src/firmware/app/control/python_controller/difference_equation_test.py]: same as above

File: src/firmware/app/control/python_controller/difference_equation_test.py
Comment (MathewMacDougall) [inline src/firmware/app/control/python_controller/difference_equation_test.py]: same as above

File: src/firmware/app/control/python_controller/difference_equation_test.py
Comment (MathewMacDougall) [inline src/firmware/app/control/python_controller/difference_equation_test.py]: same as above

File: src/firmware/app/control/python_controller/difference_equation_test.py
Comment (MathewMacDougall) [inline src/firmware/app/control/python_controller/difference_equation_test.py]: same as above

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
nit

File: src/firmware/app/control/python_controller/difference_equation_test.py
Comment (EvanMorcom) [reply src/firmware/app/control/python_controller/difference_equation_test.py]: Renamed variables and added comments to explain the variables

File: src/firmware/app/control/python_controller/difference_equation_test.py
Comment (EvanMorcom) [reply src/firmware/app/control/python_controller/difference_equation_test.py]: 👍 See above

**Review by MathewMacDougall — APPROVED** [review]
LGucciTM

**Review by jonathanlew — APPROVED** [review]
LGTM!

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Pull request merged**
